First draft of Passport Model (#27)
yaronf/draft-tls-attestation: draft-fossati-tls-attestation.md

- the Android Key Attestation, and
- Apple Key Attestation. 

Likewise, there are different encodings available for attestation results. One
such encoding, AR4SI {{?I-D.ietf-rats-ar4si}} is being standardized by the RATS
working group.

This specification defines how to support the background check model in the TLS
handshake, such that the details about the attestation technology are agnostic

> to the TLS handshake itself. Similarly, support for the passport model is

Reviewer: I would rephrase it as follows.

This specification supports both the background check and passport model, during TLS Handshake.

In background check model the details about the attestation technology are agnostice to TLS handshake itself.
Similarly, in the passport model, the details about the attestation results encoding and trust relationships are agnostic to  the TLS handshake.

defined such that details about the attestation results encodings and trust
relationships are agnostic to the TLS handshake.

To give the peer information that the handshake signing key is properly secured,
the associated attestation result has to be appraised by the peer. This must be
the case when either of the two remote attestation topologies is used. Hence,
attestation evidence about the security state of the signing key is needed,

> which is typically associated with evidence about the overall platform state.

Reviewer: Hence the attestation evidence (to be appraised by an attestation verifier )
must contain the security state of the signing key and also the security of the
overall platform state.

Reviewer 2: attestation verifier is not used otherwise in the draft, so probably skip "attestation". 

Also, slightly shorter and concise version: 

Hence, the attestation evidence (to be appraised by the verifier)
must contain the security state of the signing key as well as the overall platform.

The platform attestation service ensures that the key attestation service has

> not been tampered with. The platform attestation service issues the Platform

Reviewer: Hence the attestation evidence (to be appraised by an attestation verifier )
must contain the security state of the signing key and also the security of the 
overall platform state.

Attestation Token (PAT) and the key attestation service issues the Key
Attestation Token (KAT). The security of the protocol critically depends on the
verifiable binding between these two logically separate units of evidence.

This document does not define how different attestation technologies are
encoded. This is accomplished by companion specifications.

# Conventions and Terminology

The following terms are used in this document:

{: vspace="0"}

TLS Identity Key (TIK): : A cryptographic key used by one of the peers to
authenticate itself during the TLS handshake.

The reader is assumed to be familiar with the vocabulary and concepts defined in
{{-rats-arch}}, and those in {{-rats-kat}}.

"Remote attestation evidence" is more succintly referred to as "evidence", and

> "remote attestation results" is more succintly referred to as "results"

Reviewer 2: "results" is a very generic word and can be confusing at times. I would suggest "attestation results".

> throughout this document. "Remote attestation credentials", or "credentials", is

Reviewer 2: Similar to above, I would suggest "attestation credentials" rather than "credentials" alone.

used to refer to both evidence and results, when no distinction needs to be made
between them.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in RFC 2119 {{RFC2119}}.

# Overview

The Remote Attestation Procedures (RATS) architecture {{RFC9334}} defines two
types of interaction models for attestation, namely the passport model and the
background check model. The subsections below explain the difference in their
interactions.

As typical with new features in TLS, the client indicates support for the new
extension in the ClientHello message. The newly introduced extensions allow

> remote attestation credentials and nonces to be exchanged. The nonces are used

Reviewer 2: Why does nonce need to be mentioned separately? Is it not a part of the "remote attestation credentials"? For example, KAT includes nonce, right?

Author: I believe nonces are mentioned separately because they have to be passed along explicitly using the new extensions (for background check, at least), separate from the credentials.

Reviewer 2: Then probably define the "remote attestation credentials" such that it is clear that nonce is not a part of that.

Author: Do you mean in the `Conventions and Terminology` section or in this paragraph? I don't think nonces are usually lumped into "credentials", though perhaps this is confusing since the nonces are incorporated in credentials.

for guaranteeing freshness of the exchanged evidence when the background check
model is in use.

When either the evidence or the results extension is successfully negotiated,
the content of the corresponding Certificate message contains a payload that is
encoded based on the wrapper defined in {{I-D.ftbs-rats-msg-wrap}}.

In TLS a client has to demonstrate possession of the private key via the
CertificateVerify message, when client-based authentication is requested. The
attestation payload must contain assertions relating to the client's TLS
Identity Key (TIK), which associate the private key with the attestation
information. These assertions may come in the form of a Key Attestation Token
(KAT), or of specific claims in an attestation result document. An example of a
KAT format utilizing the EAT format can be found in {{I-D.bft-rats-kat}}.

The relying party can obtain and appraise the remote attestation results either
directly from the Certificate message (in the passport model), or by relaying

> the evidence from the Certificate message to the verifier. Subsequently, the

Reviewer: I would re-phrase it to;

 or by relaying the evidence from the Certificate message to the verifier and receiving the attestation results. Subsequently, the ...

attested key is used to verify the CertificateVerify message.

When using the passport model, the remote attestation results obtained by the
attester from its trusted verifiers can be cached and used for any number of
subsequent TLS handshakes. This allows the two sides to amortize the cost of
remote attestation by reducing the frequency with which evidence must be
produced and appraised. However, this does create a risk of stale attestation
results being used for authentication. The relying party must therefore be
empowered to reject an authentication attempt with results that do not meet its
staleness policy. This should trigger a new evidence appraisal on the attester

> side.

Reviewer: Suggested edit:
When using the passport model, the remote attestation results obtained by the
attester from its trusted verifiers can be cached and used for any number of
subsequent TLS handshakes. This allows the two sides to amortize the cost of
remote attestation by reducing the frequency with which remote attestation needs to be
performed with the verifier.  However, this does create a risk of stale attestation
results being used for authentication. The relying party must therefore be
empowered to reject an authentication attempt with results that do not meet its
acceptance policy. This should trigger a new attestation protocol between attester 
and the verifier.

Reviewer: Also a generic question, we should specify 
(a) how existing protocols built in the TLS system allow us to reject the Attestation results

(b) If a suitable cause needs to be added in the Reject Message, what the new value should be?

> When using the passport model, the remote attestation results obtained by the
> attester from its trusted verifiers can be cached and used for any number of
> subsequent TLS handshakes. This allows the two sides to amortize the cost of
> remote attestation by reducing the frequency with which evidence must be
> produced and appraised. However, this does create a risk of stale attestation
> results being used for authentication. The relying party must therefore be
> empowered to reject an authentication attempt with results that do not meet its
> staleness policy. This should trigger a new evidence appraisal on the attester
> side.

Reviewer 3: ```suggestion
When using the passport model, the remote attestation results obtained by the
attester from its trusted verifiers can be cached and used for any number of
subsequent TLS handshakes, as long as the freshness policy requirements are satisfied. ```

Author: Do I need to explain what the RP has to do to signal that these freshness requirements were not met?

Reviewer 3: From their point of view, I don't think there's anything special: it's a failure in the credential verification.

# Use of Remote Attestation credentials in the TLS handshake

For both the passport model (described in section 5.1 of {{RFC9334}}) and
background check model (described in Section 5.2 of {{RFC9334}}) the following
modes of operation are allowed when used with TLS, namely:

- TLS client is the attester, 
- TLS server is the attester, and
- TLS client and server mutually attest towards each other. 

We will show the message exchanges of the first two cases in sub-sections below.

> Mutual authentication via attestation combines these two (non-interfering)

Reviewer 2: Why not simply "Mutual attestation"?

Reviewer 4: @muhammad-usama-sardar Because from a TLS point of view this is authentication.

~~~~
{: #figure-background-check-model2 title="TLS Client Providing Evidence to TLS Server."}

> ## TLS Client authenticating using results 

Reviewer 4: ```suggestion
## TLS Client authenticating using attestation results 
```

In this use case the TLS client, as the attester, provides attestation results
to the TLS server. The TLS client is the attester and the the TLS server acts as
a relying party. Prior to delivering its Certificate message, the client must
contact the verifier (not shown in the diagram) to receive the attestation

> results that it will use as credentials.

Reviewer 4: Here again the 2nd sentence duplicates the first one.

Author: I'll likely rebase this PR on top of the early exporters one, so will take care of this afterwards.

~~~~
       Client                                           Server

Key  ^ ClientHello
Exch | + results_proposal
     | + key_share*
     | + signature_algorithms*
     v                         -------->
                                                  ServerHello  ^ Key
                                                 + key_share*  | Exch
                                                               v
                                        {EncryptedExtensions}  ^  Server
                                           + results_proposal  |  Params
                                         {CertificateRequest}  v  
                                                {Certificate}  ^
                                          {CertificateVerify}  | Auth
                                                   {Finished}  v
                               <--------  [Application Data*]
     ^ {Certificate}
Auth | {CertificateVerify}
     v {Finished}              -------->
       [Application Data]      <------->  [Application Data]
~~~~
{: #figure-passport-model1 title="TLS Client Providing Results to TLS Server."}

> ## TLS Server authenticating using results

Reviewer 4: ```suggestion
## TLS Server authenticating using attestation results
```

The hash algorithm negotiatied within the handshake must be used wherever
hashing is required for the binder.

> # Results Extensions (Passport Model)

Reviewer 4: ```suggestion
# Attestation Results Extensions (Passport Model)
```

> IANA is requested to allocate a value in the "TLS Alerts"

Reviewer 4: IANA: I think we're also adding two more extensions, right?
